tests/helpers.py

```diff
@@ -0,0 +1,389 @@
+        return
+
+    try:
+        url = 'http://ssb.stsci.edu/cdbs/{}/{}'.format(refdir, refname)
+        _download_file(url, refname, timeout=timeout)
+    except Exception:  # Fall back to FTP
+        url = 'ftp://ftp.stsci.edu/cdbs/{}/{}'.format(refdir, refname)
+        _download_file(url, refname, timeout=timeout)
+
+
+def download_file_cgi(tree, project, filename, filemode='wb', timeout=30,
+                      allow_remote_ref=False):
+    """
+    Download remote data to current directory from Central Storage
+    or using CGI script interface.
+
+    Parameters
+    ----------
+    tree : {'rt', 'rtx', 'null'}
+        Test tree:
+
+        * rt = dev
+        * rtx = public
+        * null = neither, grab first match (e.g., for STAK)
+
+    project : str
+        Path containing data in the test tree.
+        For example, ``hstcal/acs/calacs_e``.
+
+    filename : str
+        Filename of the data.
+
+    filemode : str
+        Mode of the output file writer. By default, it saves as binary
+        and overwrites existing file.
+
+    timeout : number or `None`
+        This is not a time limit on the entire response download;
+        rather, an exception is raised if the server has not issued
+        a response for timeout seconds (more precisely, if no bytes
+        have been received on the underlying socket for timeout seconds).
+        If no timeout is specified explicitly, requests do not time out.
+
+    allow_remote_ref : bool
+        If set to `True`, instead of copying or downloading to current
+        directory, return the full path. This should only be used for
+        "truth" images.
+
+    """
+    # NOTE: This could be explicitly controlled using pytest fixture
+    #       but too many ways to do the same thing would be confusing.
+    #       Refine this logic if using pytest fixture.
+    cs_file = os.path.join('/eng/ssb2/tests', tree, project, filename)
+
+    if os.path.isfile(cs_file):
+        if allow_remote_ref:
+            return cs_file
+        shutil.copyfile(cs_file, filename)
+    else:
+        url = ('http://ssb.stsci.edu/cgi-bin/remote_testing.cgi?'
+               'tree={}&project={}&name={}'.format(tree, project, filename))
+        if allow_remote_ref:
+            return url
+        _download_file(url, filename, filemode=filemode, timeout=timeout)
+
+
+def _get_reffile(hdr, key):
+    """Get ref file from given key in given FITS header."""
+    ref_file = None
+    if key in hdr:  # Keyword might not exist
+        ref_file = hdr[key].strip()
+        if ref_file.upper() == 'N/A':  # Not all ref file is defined
+            ref_file = None
+    return ref_file
+
+
+def ref_from_image(input_image):
+    """
+    Return a list of reference filenames, as defined in the primary
+    header of the given input image, necessary for calibration; i.e.,
+    only those associated with ``*CORR`` set to ``PERFORM`` will be
+    considered.
+    """
+    # NOTE: Add additional mapping as needed.
```

Comment:
Can this be replaced by https://github.com/spacetelescope/crds/blob/master/crds/bestrefs/bestrefs.py?

Reply:
This one in helpers is simply to make sure the necessary *keywords* are present. It does not impose any *value* requirements.

Whether to use latest CRDS *value* mapping or not is a separate discussion. Ideally, we need to be able to tell when there is a diff, whether it is caused by code change or CRDS change (we might or might not care about the latter).

Reply:
@stscicrawford , let's table this one for "future work" (if necessary). Current `/grp/hst/cdbs` look-up works fine and does not introduce extra `crds` package dependency.